README
#
# Copyright 2005, 2006 Free Software Foundation, Inc.
# 
# This file is part of GNU Radio
# 
# GNU Radio is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
# 
# GNU Radio is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with GNU Radio; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street,
# Boston, MA 02110-1301, USA.
# 


What is this?
--------------------

This gnuradio module is a simple DCF77 transmitter or to be more precise
it is a generator for a time signal, that is usually transmitted by the
DCF77 station.

To learn more about the DCF77 station, please read this wikipedia page:
https://secure.wikimedia.org/wikipedia/en/wiki/DCF77

The time signal generated by this gnuradio module can be used to set the
time of radio controlled clocks. In two of three cases you can set the time
for a powered up clock in three to five minutes. If your clock is already
synced with a time signal, it highly depends on the clock, when it tries
to synchronise again. Some clocks will only sync once a day.

Installation
--------------------

This package requires that gnuradio-core is already installed. It
also depends on some GNU Radio prerequisites, such as boost.

To build the code from the tarball use the normal recipe:

  $ ./configure
  $ make
  $ make check

If you're building from CVS, you'll need to use this sequence, since
CVS doesn't contain configure or the generated Makefiles.

  $ ./bootstrap
  $ ./configure
  $ make
  $ make check


The doc directory is not built by default.  This is to avoid spurious
build problems on systems that don't have xmlto installed.  If you
have xmlto and its dependencies installed, you can build the html
version of the howto article by cd'ing to doc and invoking make.

Usage
--------------------

Just have a look into start.sh and dcf77_tx_test.py prepare your USRP
and run start.sh.

Plug in the LFTX module in order to prepare your USRP. For testing
you can connect the LFTX to a coil and place the clock in the
center of the coil. You don't need any amplifier for this setup.


Legal aspects
--------------------

Obviously there are legal aspects. Transmitting time signals outside
of special radio test rooms is not allowed in most cases.


Author
--------------------

Martin Schobert <martin@weltregierung.de>



ChangeLog
2007-01-09  Foo Bar

	* made release 0.1

#
# Copyright 2005 Free Software Foundation, Inc.
# 
# This file is part of GNU Radio
# 
# GNU Radio is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
# 
# GNU Radio is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with GNU Radio; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street,
# Boston, MA 02110-1301, USA.
# 


src/python/dcf77_tx_test.py
#!/usr/bin/env python

"""
Simple DCF77 Transmitter test
"""

from gnuradio import gr, eng_notation
from gnuradio import usrp
from gnuradio import audio
from gnuradio import blks
from gnuradio.eng_option import eng_option
from optparse import OptionParser
from usrpm import usrp_dbid
import sys
from gnuradio import dcf77

from gnuradio.wxgui import stdgui, fftsink, scopesink
from gnuradio import tx_debug_gui
import wx

class pipeline(gr.hier_block):
    def __init__(self, fg,  lo_freq, usrp_rate):


        cl = dcf77.clock(07, # hour
                         35-2, # min
                         6,  # day in month
                         7,  # day of week
                         2,  # month
                         8,  # year
                         0)  # loop
        tx = dcf77.mod(usrp_rate)
	f2c = gr.float_to_complex()

        mixer = gr.multiply_ff()

        fg.connect (cl, tx, f2c)
        
        gr.hier_block.__init__(self, fg, cl, f2c)


class tx_graph (stdgui.gui_flow_graph):
    def __init__(self, frame, panel, vbox, argv):

        stdgui.gui_flow_graph.__init__ (self, frame, panel, vbox, argv)

        parser = OptionParser (option_class=eng_option)
        parser.add_option("-T", "--tx-subdev-spec", type="subdev", default=None,
                          help="select USRP Tx side A or B")
        parser.add_option("-f", "--freq", type="eng_float", default=None,
                           help="set Tx frequency to FREQ [required]", metavar="FREQ")
        parser.add_option("","--debug", action="store_true", default=False,
                          help="Launch Tx debugger")
        (options, args) = parser.parse_args ()

        if len(args) != 0:
            parser.print_help()
            sys.exit(1)

        if options.freq is None:
            sys.stderr.write("usrp_tx_video.py: must specify frequency with -f FREQ\n")
            parser.print_help()
            sys.exit(1)

        # ----------------------------------------------------------------

        self.u = usrp.sink_c ()       # the USRP sink (consumes samples)

        self.dac_rate = self.u.dac_rate()                    # 128 MS/s
        self.usrp_interp = 128
        self.usrp_rate = self.dac_rate / self.usrp_interp


        self.u.set_interp_rate(self.usrp_interp)

        # determine the daughterboard subdevice we're using
        if options.tx_subdev_spec is None:
            options.tx_subdev_spec = usrp.pick_tx_subdevice(self.u)

        m = usrp.determine_tx_mux_value(self.u, options.tx_subdev_spec)
        self.u.set_mux(m)
        self.subdev = usrp.selected_subdev(self.u, options.tx_subdev_spec)
        print "Using TX d'board %s with rate %ld" % (self.subdev.side_and_name(), self.usrp_rate)

        self.subdev.set_gain(self.subdev.gain_range()[1])    # set max Tx gain
	self.u.tune(self.subdev._which, self.subdev, options.freq)
        self.subdev.set_enable(True)                         # enable transmitter

        t = pipeline(self, options.freq, self.usrp_rate)

        gain = gr.multiply_const_cc (32000.0)

        # connect it all
        self.connect (t, gain)
        self.connect (gain, self.u)

        # plot an FFT to verify we are sending what we want
        if 1:
            post_mod_fft = fftsink.fft_sink_c(self, panel, title="Signal in frequency domain",
                                              fft_size=64, sample_rate=self.usrp_rate,
                                              y_per_div=20, ref_level=40)

            #post_mod_oscope = scopesink.scope_sink_c(self, panel, sample_rate=self.usrp_rate,
            #                                         title="Signal in time domain", v_scale = 1,
            #                                         t_scale=0.0001)

            self.connect (t, post_mod_fft)
            #self.connect (t, post_mod_oscope)
            vbox.Add (post_mod_fft.win, 1, wx.EXPAND)
            #vbox.Add (post_mod_oscope.win, 1, wx.EXPAND)
            

        if options.debug:
            self.debugger = tx_debug_gui.tx_debug_gui(self.subdev)
            self.debugger.Show(True)


def main ():
    app = stdgui.stdapp (tx_graph, "DCF77 transmitter")
    app.MainLoop ()

if __name__ == '__main__':
    main ()


config.h
/* config.h.  Generated by configure.  */
/* config.h.in.  Generated from configure.ac by autoheader.  */

/* Define if for loop scoping is broken */
#define CXX_HAS_BUGGY_FOR_LOOPS 

/* Define if C++ is missing bool type */
/* #undef CXX_HAS_NO_BOOL */

/* Define if has std::isnan */
/* #undef CXX_HAS_STD_ISNAN */

/* Define to 1 if you have the `clock_gettime' function. */
#define HAVE_CLOCK_GETTIME 1

/* Define to 1 if you have the <dlfcn.h> header file. */
#define HAVE_DLFCN_H 1

/* Define to 1 if you have the <fcntl.h> header file. */
#define HAVE_FCNTL_H 1

/* Define to 1 if you have the `getopt' function. */
#define HAVE_GETOPT 1

/* Define to 1 if you have the `gettimeofday' function. */
#define HAVE_GETTIMEOFDAY 1

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <io.h> header file. */
/* #undef HAVE_IO_H */

/* Define to 1 if you have the <limits.h> header file. */
#define HAVE_LIMITS_H 1

/* Define to 1 if you have the <memory.h> header file. */
#define HAVE_MEMORY_H 1

/* Define to 1 if you have the `nanosleep' function. */
#define HAVE_NANOSLEEP 1

/* Define if you have POSIX threads libraries and header files. */
/* #undef HAVE_PTHREAD */

/* Define to 1 if you have the <Python.h> header file. */
#define HAVE_PYTHON_H 1

/* Define to 1 if you have the `rand' function. */
#define HAVE_RAND 1

/* Define to 1 if you have the `random' function. */
#define HAVE_RANDOM 1

/* Define to 1 if you have the `sigaction' function. */
#define HAVE_SIGACTION 1

/* Define to 1 if you have the `sleep' function. */
#define HAVE_SLEEP 1

/* Define to 1 if you have the `srand' function. */
#define HAVE_SRAND 1

/* Define to 1 if you have the `srandom' function. */
#define HAVE_SRANDOM 1

/* Define to 1 if the system has the type `ssize_t'. */
#define HAVE_SSIZE_T 1

/* Define to 1 if you have win32 Sleep */
/* #undef HAVE_SSLEEP */

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if the system has the type `struct timespec'. */
#define HAVE_STRUCT_TIMESPEC 1

/* Define to 1 if the system has the type `struct timezone'. */
#define HAVE_STRUCT_TIMEZONE 1

/* Define to 1 if you have the <sys/ioctl.h> header file. */
#define HAVE_SYS_IOCTL_H 1

/* Define to 1 if you have the <sys/mman.h> header file. */
#define HAVE_SYS_MMAN_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/time.h> header file. */
#define HAVE_SYS_TIME_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have <sys/wait.h> that is POSIX.1 compatible. */
#define HAVE_SYS_WAIT_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if you have the `usleep' function. */
#define HAVE_USLEEP 1

/* Define to 1 if you have the <winbase.h> header file. */
/* #undef HAVE_WINBASE_H */

/* Define to 1 if you have the <windows.h> header file. */
/* #undef HAVE_WINDOWS_H */

/* Define to 1 if you have the <winioctl.h> header file. */
/* #undef HAVE_WINIOCTL_H */

/* Define if mkdir accepts only one arg */
/* #undef MKDIR_TAKES_ONE_ARG */

/* Define to disable asserts (don't doit!) */
/* #undef NDEBUG */

/* Define to 1 to enable NT thread */
/* #undef OMNITHREAD_NT */

/* Define to 1 to enable pthread */
#define OMNITHREAD_POSIX 1

/* Name of package */
#define PACKAGE "gr-dcf77-transmitter"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME ""

/* Define to the full name and version of this package. */
#define PACKAGE_STRING ""

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME ""

/* Define to the version of this package. */
#define PACKAGE_VERSION ""

/* Define to the necessary symbol if this constant uses a non-standard name on
   your system. */
/* #undef PTHREAD_CREATE_JOINABLE */

/* Define to 1 if you have the ANSI C header files. */
#define STDC_HEADERS 1

/* Define to 1 if you can safely include both <sys/time.h> and <time.h>. */
#define TIME_WITH_SYS_TIME 1

/* Version number of package */
#define VERSION "3.0svn"

/* Define to empty if `const' does not conform to ANSI C. */
/* #undef const */

/* Define to `__inline__' or `__inline' if that's what the C compiler
   calls it, or to nothing if 'inline' is not supported under any name.  */
#ifndef __cplusplus
/* #undef inline */
#endif

/* Define to `unsigned' if <sys/types.h> does not define. */
/* #undef size_t */

// Workaround for compilers with buggy for-loop scoping
// That's quite a few compilers actually including recent versions of
// Dec Alpha cxx, HP-UX CC and SGI CC.
// The trivial "if" statement provides the correct scoping to the
// for loop

#ifdef CXX_HAS_BUGGY_FOR_LOOPS
/* #undef for */
#define for if(1) for
#endif


// If the C++ compiler we use doesn't have bool, then
// the following is a near-perfect work-around.
// You must make sure your code does not depend on "int" and "bool"
// being two different types, in overloading for instance.

#ifdef CXX_HAS_NO_BOOL
#define bool int
#define true 1
#define false 0
#endif



/* Define missing prototypes, implemented in replacement lib */
#ifdef  __cplusplus
extern "C" {
#endif

#ifndef HAVE_GETOPT
int getopt (int argc, char * const argv[], const char * optstring);
extern char * optarg;
extern int optind, opterr, optopt;
#endif

#ifndef HAVE_USLEEP
int usleep(unsigned long usec);	/* SUSv2 */
#endif

#ifndef HAVE_NANOSLEEP
#ifndef HAVE_STRUCT_TIMESPEC
#if HAVE_SYS_TYPES_H
# include <sys/types.h>	/* need time_t */
#endif
struct timespec {
	time_t	tv_sec;
	long	tv_nsec;
};
#endif
static inline int nanosleep(const struct timespec *req, struct timespec *rem) { return usleep(req->tv_sec*1000000+req->tv_nsec/1000); }
#endif

#if defined(HAVE_SSLEEP) && !defined(HAVE_SLEEP)
#ifdef HAVE_WINBASE_H
#include <windows.h>
#include <winbase.h>
#endif
/* TODO: what about SleepEx? */
static inline unsigned int sleep (unsigned int nb_sec) { Sleep(nb_sec*1000); return 0; }
#endif

#ifndef HAVE_GETTIMEOFDAY
#ifdef HAVE_SYS_TIME_H
#include <sys/time.h>
#endif
#ifndef HAVE_STRUCT_TIMEZONE
struct timezone {
        int  tz_minuteswest;
	int  tz_dsttime;
};
#endif
int gettimeofday(struct timeval *tv, struct timezone *tz);
#endif

#if !defined(HAVE_RANDOM) && defined(HAVE_RAND)
#include <stdlib.h>
static inline long int random (void) { return rand(); }
#endif

#if !defined(HAVE_SRANDOM) && defined(HAVE_SRAND)
static inline void srandom (unsigned int seed) { srand(seed); }
#endif

#ifndef HAVE_SSIZE_T
typedef size_t ssize_t;
#endif

#ifdef  __cplusplus
}
#endif



src/lib/dcf77_mod.h
/* -*- c++ -*- */
/*
 * Copyright 2004 Free Software Foundation, Inc.
 * 
 * This file is part of GNU Radio
 * 
 * GNU Radio is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 * 
 * GNU Radio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with GNU Radio; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street,
 * Boston, MA 02110-1301, USA.
 */
#ifndef INCLUDED_DCF77_MOD_H
#define INCLUDED_DCF77_MOD_H

#include <gr_block.h>

class dcf77_mod;

/*
 * We use boost::shared_ptr's instead of raw pointers for all access
 * to gr_blocks (and many other data structures).  The shared_ptr gets
 * us transparent reference counting, which greatly simplifies storage
 * management issues.  This is especially helpful in our hybrid
 * C++ / Python system.
 *
 * See http://www.boost.org/libs/smart_ptr/smart_ptr.htm
 *
 * As a convention, the _sptr suffix indicates a boost::shared_ptr
 */
typedef boost::shared_ptr<dcf77_mod> dcf77_mod_sptr;

/*!
 * \brief Return a shared_ptr to a new instance of dcf77_mod.
 *
 * To avoid accidental use of raw pointers, dcf77_mod's
 * constructor is private.  dcf77_make_mod is the public
 * interface for creating new instances.
 */
dcf77_mod_sptr dcf77_make_mod (long sample_rate);

/*!
 * \brief generate AM modulated carrier signal from bit input stream
 * \ingroup block
 *
 */

const int STATE_AMPL_SEND_0_low = 0;
const int STATE_AMPL_SEND_0_high = 1;
const int STATE_AMPL_SEND_1_low = 2;
const int STATE_AMPL_SEND_1_high = 3;
const int STATE_AMPL_NO_MOD = 4;

class dcf77_mod : public gr_block
{
private:


  // The friend declaration allows dcf77_make_mod to
  // access the private constructor.

  friend dcf77_mod_sptr dcf77_make_mod (long sample_rate);

  dcf77_mod (long sample_rate);	// private constructor

  bool state_change;
  long sample_rate;
  //long samples_produced;
  int state;
  float ampl;
  int next_state;
  long hold_state_for_n_samples;

 public:
  ~dcf77_mod ();	// public destructor

  void forecast(int noutput_items, gr_vector_int &ninput_items_required);

  // Where all the action really happens

  int general_work (int noutput_items,
		    gr_vector_int &ninput_items,
		    gr_vector_const_void_star &input_items,
		    gr_vector_void_star &output_items);
};

#endif /* INCLUDED_DCF77_MOD_H */


src/lib/dcf77_clock.h
/* -*- c++ -*- */
/*
 * Copyright 2004 Free Software Foundation, Inc.
 * 
 * This file is part of GNU Radio
 * 
 * GNU Radio is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 * 
 * GNU Radio is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with GNU Radio; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street,
 * Boston, MA 02110-1301, USA.
 */
#ifndef INCLUDED_DCF77_CLOCK_H
#define INCLUDED_DCF77_CLOCK_H

#include <gr_block.h>

class dcf77_clock;

/*
 * We use boost::shared_ptr's instead of raw pointers for all access
 * to gr_blocks (and many other data structures).  The shared_ptr gets
 * us transparent reference counting, which greatly simplifies storage
 * management issues.  This is especially helpful in our hybrid
 * C++ / Python system.
 *
 * See http://www.boost.org/libs/smart_ptr/smart_ptr.htm
 *
 * As a convention, the _sptr suffix indicates a boost::shared_ptr
 */
typedef boost::shared_ptr<dcf77_clock> dcf77_clock_sptr;

/*!
 * \brief Return a shared_ptr to a new instance of dcf77_clock.
 *
 * To avoid accidental use of raw pointers, dcf77_clock's
 * constructor is private.  dcf77_make_clock is the public
 * interface for creating new instances.
 */

dcf77_clock_sptr dcf77_make_clock(int hour, int min, int mday, int dow, 
				  int month, int year, int loops);

/*!
 * \brief generate AM modulated carrier signal from bit input stream
 * \ingroup block
 *
 */

class dcf77_clock : public gr_block
{
private:


  // The friend declaration allows dcf77_make_clock to
  // access the private constructor.

  friend dcf77_clock_sptr dcf77_make_clock (int hour, int min, int mday, int dow, 
					    int month, int year, int loops);

  dcf77_clock (int hour, int min, int mday, int dow, 
	       int month, int year, int loops); // private constructor

  int hour, min, mday, dow, month, year;
  int bit_buf[60];
  int bit_buf_pos;

  int loops;
  int loop_cnt;

  void fill_bit_buffer();
  int calc_bit_vector(int value, int * valency, int len, int * buf_ptr);

 public:
  ~dcf77_clock ();	// public destructor

  void forecast(int noutput_items, gr_vector_int &ninput_items_required);

  // Where all the action really happens

  int general_work (int noutput_items,
		    gr_vector_int &ninput_items,
		    gr_vector_const_void_star &input_items,
		    gr_vector_void_star &output_items);
};

#endif /* INCLUDED_DCF77_CLOCK_H */


src/python/qa_dcf77.py
#!/usr/bin/env python
#
# Copyright 2004 Free Software Foundation, Inc.
# 
# This file is part of GNU Radio
# 
# GNU Radio is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
# 
# GNU Radio is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with GNU Radio; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 51 Franklin Street,
# Boston, MA 02110-1301, USA.
# 

from gnuradio import gr, gr_unittest
import dcf77

class qa_dcf77 (gr_unittest.TestCase):

    def setUp (self):
        self.fg = gr.flow_graph ()

    def tearDown (self):
        self.fg = None

    def test_001_clock (self):
        cl = dcf77.clock(14, 44, 1, 7, 1, 6, 3)

        dst = gr.vector_sink_i()
        self.fg.connect (cl, dst)

        self.fg.run ()
        result_data = dst.data ()

        expected_result = (
            0,                            # 0: always 0
            0,0,0,0,0,0,0,0,0,0,0,0,0,0,  # 1-14: reserved/ unused
            0,                            # 15: 0 =default antenna
            0,                            # 16: 0 = no mez/mesz-change
            0,1,                          # 17-18: 01= mez
            0,                            # 19: no leap second with next 59 sec
            1,                            # 20: start bit for time block, always 1
            0,0,1,0,0,0,1,                # 21-27: min = 1, 2, 4, 8, 10, 20, 40 -- 44
            0,                            # 28: check bit for minutes: 0 = even, 1 = odd
            0,0,1,0,1,0,                  # 29-34: hours
            0,                            # 35: check bit
            1,0,0,0,0,0,                  # 36-41: day within month
            1,1,1,                        # 42-44: day of week (1-7)
            1,0,0,0,0,                    # 45-49: month
            0,1,1,0,0,0,0,0,              # 50-57: year
            1,                            # 58: check bit, here 1
            2,
            0,                            # 0: always 0
            0,0,0,0,0,0,0,0,0,0,0,0,0,0,  # 1-14: reserved/ unused
            0,                            # 15: 0 =default antenna
            0,                            # 16: 0 = no mez/mesz-change
            0,1,                          # 17-18: 01= mez
            0,                            # 19: no leap second with next 59 sec
            1,                            # 20: start bit for time block, always 1
            1,0,1,0,0,0,1,                # 21-27: min = 1, 2, 4, 8, 10, 20, 40 -- 45
            1,                            # 28: check bit for minutes: 0 = even, 1 = odd
            0,0,1,0,1,0,                  # 29-34: hours
            0,                            # 35: check bit
            1,0,0,0,0,0,                  # 36-41: day within month
            1,1,1,                        # 42-44: day of week (1-7)
            1,0,0,0,0,                    # 45-49: month
            0,1,1,0,0,0,0,0,              # 50-57: year
            1,                            # 58: check bit, here 1
            2,
            0,                            # 0: always 0
            0,0,0,0,0,0,0,0,0,0,0,0,0,0,  # 1-14: reserved/ unused
            0,                            # 15: 0 =default antenna
            0,                            # 16: 0 = no mez/mesz-change
            0,1,                          # 17-18: 01= mez
            0,                            # 19: no leap second with next 59 sec
            1,                            # 20: start bit for time block, always 1
            0,1,1,0,0,0,1,                # 21-27: min = 1, 2, 4, 8, 10, 20, 40 -- 46
            1,                            # 28: check bit for minutes: 0 = even, 1 = odd
            0,0,1,0,1,0,                  # 29-34: hours
            0,                            # 35: check bit
            1,0,0,0,0,0,                  # 36-41: day within month
            1,1,1,                        # 42-44: day of week (1-7)
            1,0,0,0,0,                    # 45-49: month
            0,1,1,0,0,0,0,0,              # 50-57: year
            1,                            # 58: check bit, here 1
            2
            )

        print "bit pos | expected | result  | error"

        for i in range(len(result_data)):
            comment = ""
            if expected_result[i] != result_data[i]:
                comment = "error"
            print "%03d     | %d        | %d       | %s" % (i,
                  expected_result[i], result_data[i], comment)
            
        self.assertFloatTuplesAlmostEqual (expected_result, result_data, len(expected_result))

    def test_002_clock_to_tx (self):
        sample_rate = 10
        expected_result = [0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           0.25, 0.25, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
                           1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]

        cl = dcf77.clock(14, 44, 1, 7, 1, 6, 2)
        tx = dcf77.mod(sample_rate)

        dst = gr.vector_sink_f ()
        self.fg.connect (cl, tx, dst)

        self.fg.run ()
        result_data = dst.data()[0:len(expected_result)]
        
        self.assertFloatTuplesAlmostEqual( expected_result, result_data,
                                           len(expected_result))


if __name__ == '__main__':
    gr_unittest.main ()
